c_src/srtp/srtp.c

```diff
@@ -0,0 +1,363 @@
+    }
+  }
+
+  // Single master key scenario
+  if (keys_length == 1 && keys_mkis_length == 0) {
+    policy->key = keys[0]->data;
+    return true;
+  }
+
+  if (keys_length != keys_mkis_length) {
+    *err = unifex_raise_args_error(env, "keys_mkis",
+                                   "must be of same length as keys");
+    return false;
+  }
+
+  policy->keys = unifex_alloc(sizeof(srtp_master_key_t *) * keys_length);
+  policy->num_master_keys = keys_length;
+  if (!policy->keys) {
+    free_master_keys_array(policy);
+    *err = unifex_raise(env, "not enough memory");
+    return false;
+  }
+
+  // Set all pointers to NULL
+  memset(policy->keys, 0, sizeof(srtp_master_key_t *) * keys_length);
+
+  for (size_t i = 0; i < keys_length; i++) {
+    srtp_master_key_t *key = unifex_alloc(sizeof(srtp_master_key_t));
+    if (!key) {
+      free_master_keys_array(policy);
+      *err = unifex_raise(env, "not enough memory");
+      return false;
+    }
+
+    key->key = keys[i]->data;
+    key->mki_id = keys_mkis[i]->data;
+    key->mki_size = keys_mkis[i]->size;
+
+    policy->keys[i] = key;
+  }
+
+  return true;
+}
+
+UNIFEX_TERM add_stream(UnifexEnv *env, UnifexState *state, int ssrc_type,
+                       uint ssrc, UnifexPayload **keys,
+                       unsigned int keys_length, UnifexPayload **keys_mkis,
+                       unsigned int keys_mkis_length, char *rtp_crypto_profile,
+                       char *rtcp_crypto_profile, uint window_size,
+                       int allow_repeat_tx) {
+  int err;
+  srtp_err_status_t serr;
+  UNIFEX_TERM uerr;
+
+  srtp_policy_t policy;
+  memset(&policy, 0, sizeof(srtp_policy_t));
+
+  policy.window_size = window_size;
+  policy.allow_repeat_tx = allow_repeat_tx;
+
+  err = unmarshal_ssrc(ssrc_type, ssrc, &policy.ssrc);
+  if (!err) {
+    return unifex_raise_args_error(env, "ssrc", "invalid");
+  }
+
+  err = set_crypto_policy_from_crypto_profile_atom(rtp_crypto_profile,
+                                                   &policy.rtp);
+  if (!err) {
+    return unifex_raise_args_error(env, "rtp", "invalid");
+  }
+
+  err = set_crypto_policy_from_crypto_profile_atom(rtcp_crypto_profile,
+                                                   &policy.rtcp);
+  if (!err) {
+    return unifex_raise_args_error(env, "rtcp", "invalid");
+  }
+
+  err = create_master_keys_array(env, keys, keys_length, keys_mkis,
+                                 keys_mkis_length, &policy, &uerr);
+  if (!err) {
+    return uerr;
+  }
+
+  serr = srtp_add_stream(state->session, &policy);
+
+  free_master_keys_array(&policy);
+
+  if (serr) {
+    return unifex_raise(env, srtp_strerror(serr));
+  }
+
+  return add_stream_result_ok(env);
+}
+
+UNIFEX_TERM remove_stream(UnifexEnv *env, UnifexState *state, uint ssrc) {
+  // htonl below is required due to legacy reasons, see
+  // https://github.com/cisco/libsrtp/issues/306
+  srtp_err_status_t serr = srtp_remove_stream(state->session, htonl(ssrc));
```

Comment:
quick though, but I am not sure if htonl is a portable function (i.e. works on Windows)

Reply:
I think so https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-htonl

Reply:
note it lies in `winsock.h`

Reply:
wouldn't care about that for now

Reply:
but if you have a better candidate, I'm all ears ;)

Reply:
https://stackoverflow.com/a/57872574